### PR TITLE
msg/async/frames_v2: optimize ceph_msg_header2 decoding

### PR DESCRIPTION
**Pull request by orozery**
Current implementation uses bufferlist rebuild,
which is expensive due to heap re-allocation of the header.
Switch to using stack storage instead.

Profiling before and after, showing the break-down of `ProtocolV2::handle_message` (call to bufferlist::c_str() is saved):
image

### CODE REVIEW

File: src/msg/async/frames_v2.h
Comment (tchaikov) [inline src/msg/async/frames_v2.h:811]: but this also incurs memcpy(). see the implementation of `decode_raw()`. but yeah, it helps to avoid rebuilding the whole chunk of `segments[SegmentIndex::Msg::HEADER]`. if its size is much greater than that of `ceph_msg_header2`, it's a great performance improvement. otherwise, i am not quite sure.

Comment (orozery) [reply src/msg/async/frames_v2.h:811]: Both implementations incur a memcpy of sizeof(ceph_msg_header2).
The difference is that the current solution also adds a call to malloc(sizeof(ceph_msg_header2)).

Comment (tchaikov) [reply src/msg/async/frames_v2.h:811]: makes sense! in this case, since the guaranteed copy elision kicks in, the caller is responsible to allocate the memory chunk (on stack) and leave it to the callee to materialize it. this save us a malloc() call.

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
```
[ RUN      ] Queue.SpawnAsyncRequest
../src/test/rgw/test_rgw_dmclock_scheduler.cc:425: Failure
Value of: context.stopped()
  Actual: false
Expected: true
[  FAILED  ] Queue.SpawnAsyncRequest (0 ms)
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by orozery** [comment]
jenkins test make check

**Review by idryomov — COMMENTED** [review]
The same (`c_str()` + `reinterpret_cast`) is done for `preamble_block_t` and all four epilogue variants.  It's possible that `c_str()` doesn't end up rebuilding in those cases so it didn't show up in your profiles but that is just by luck.  Should we wire up the raw encoder for those five structs too?

**Comment by orozery** [comment]
> The same (`c_str()` + `reinterpret_cast`) is done for `preamble_block_t` and all four epilogue variants. It's possible that `c_str()` doesn't end up rebuilding in those cases so it didn't show up in your profiles but that is just by luck. Should we wire up the raw encoder for those five structs too?

It indeed does not show up, but it's not by luck.
The preamble is decoded in from a single contiguous buffer, created by:
```
READ(rx_frame_asm.get_preamble_onwire_len(),
              handle_read_frame_preamble_main)
```

Same for the epilogue.

The `ceph_msg_header2` on the other hand, the bufferlist comes from the segment data, which is also contiguous, but splicing out the segment crc causes the bufferlist to be composed of two buffers, the other one being an empty carriage.
So `ceph_msg_header2` always have to be rebuilt, while the preamble and the epilogue do not.

**Comment by idryomov** [comment]
I'd still consider it by luck.  What if `READ` or something else on the read path changes and the buffer is no longer contiguous?

It seems that the raw encoder is strictly better than `c_str()` + `reinterpret_cast` or am I missing something?

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-07-13_00:59:36-rados-wip-kefu-testing-2021-07-12-2312-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-13_14:42:24-rados-wip-kefu-testing-2021-07-12-2312-distro-basic-smithi/

**Comment by tchaikov** [comment]
> Should we wire up the raw encoder for those five structs too?

@idryomov Hi Ilya, do we need to wait for the settlement of the discussion on further improvement in this direction or we can merge this change first?

**Comment by tchaikov** [comment]
> It seems that the raw encoder is strictly better than `c_str()` + `reinterpret_cast` or am I missing something?

i think the combination of copy elision + raw encoder is strictly better than `c_str()` + `reinterpret_cast`.

**Comment by orozery** [comment]
> I'd still consider it by luck. What if `READ` or something else on the read path changes and the buffer is no longer contiguous?
> 
> It seems that the raw encoder is strictly better than `c_str()` + `reinterpret_cast` or am I missing something?

Actually, I just went over the code that parses `preamble_block_t` and the epilogues, and it seems that it currently costs nothing, since it only calls `c_str()`, which simply returns a pointer to existing buffer, and then `reinterpret_cast`. i.e there is no copy whatsoever. 
On the other hand, if we change it to use a decoder, it will copy the data from the bufferlist to a new struct allocated on stack.

File: src/msg/async/frames_v2.h
Comment (idryomov) [inline src/msg/async/frames_v2.h:807]: Since it is no longer a reference, `const` is useless and `inline` has always been redundant:
```suggestion
  ceph_msg_header2 header() {
```

File: src/msg/async/frames_v2.h
Comment (idryomov) [inline src/msg/async/frames_v2.h:809]: Get rid of `hdrbl`:
```suggestion
    auto it = segments[SegmentIndex::Msg::HEADER].cbegin();
```

**Review by idryomov — COMMENTED** [review]
OK, let's not touch the preamble and the epilogues then.  A couple of nits:

**Comment by idryomov** [comment]
I wonder if either `splice()` or `rebuild()` should be made smarter.  From a `bufferlist` user POV, a splice from the end of the contiguous buffer leading to `c_str()` ending up in `rebuild()` and actually reallocating is counter-intuitive.

**Comment by idryomov** [comment]
@rzarzynski any thoughts on ^^^?

**Comment by orozery** [comment]
> I wonder if either `splice()` or `rebuild()` should be made smarter. From a `bufferlist` user POV, a splice from the end of the contiguous buffer leading to `c_str()` ending up in `rebuild()` and actually reallocating is counter-intuitive.

Good point.
I definitely think that `c_str()` can be made a little smarter to detect a case more cases that do not require rebuild.
Given that, perhaps it's better approach to always prefer `reinterpret_cast`, and use `ceph::decode` only as a fallback.
In that case, this commit will be reverted.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-07-17_04:37:01-rados-wip-kefu-testing-2021-07-17-1006-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-17_15:53:41-rados-wip-kefu-testing-2021-07-17-1518-distro-basic-smithi/

one of the failures is tracked by:

- https://tracker.ceph.com/issues/50245

the e2e failures are not related.

**Comment by idryomov** [comment]
I filed https://tracker.ceph.com/issues/51725 for `c_str()` improvements but until that happens we definitely want this in as it saves a heap allocation on the hot path.

Please fold the outstanding nits and I'll merge.

**Review by rzarzynski — APPROVED** [review]
Apologizes for the late review. LGTM after fixing the nits.

Thanks for pointing this out, @orozery!

**Comment by rzarzynski** [comment]
WIP branch with `_carriage` optimization in `c_str()` (untested!): https://github.com/ceph/ceph/compare/master...rzarzynski:wip-bl-optimize-emptybptrs.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-07-21_08:19:42-rados-wip-kefu-testing-2021-07-21-1257-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-21_13:52:36-rados-wip-kefu-testing-2021-07-21-2006-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-21_14:42:35-rados-wip-kefu-testing-2021-07-21-2006-distro-basic-smithi/

e2e failures tracked by https://tracker.ceph.com/issues/51728

**Comment by tchaikov** [comment]
the `c_str()` optimization is created at #42417 .

**Comment by tchaikov** [comment]
@orozery hi Or, would you mind addressing the nits ?

**Comment by orozery** [comment]
@tchaikov sure, I will fix the small nits on my open PRs (including this) ASAP.
I'm currently on PTO with no work access. I will get to this first thing when I'm back.
Hopefully next week.

**Comment by tchaikov** [comment]
@orozery thank you, and enjoy your time off =)

**Comment by idryomov** [comment]
> Given that, perhaps it's a better approach to always prefer reinterpret_cast, and use ceph::decode only as a fallback.
> In that case, this commit will be reverted.

I'm not sure we need this now given Radek's https://github.com/ceph/ceph/pull/42417.

**Comment by orozery** [comment]
Verified that given #42417 this PR is no longer needed.

**Pull request closed without merging**
